debugging-modules/README.md

```diff
@@ -0,0 +1,264 @@
+about a WebAssembly module. A WebAssembly runtime can leverage these interfaces
+in its developer tools to translate from Wasm-level debugging information into a
+user-friendly, source-level presentation.
+
+> **Note:** this is an incomplete, work-in-progress, strawperson proposal!
+> Nothing here is set in stone and everything is likely to change!
+
+## Table of Contents
+
+- [Motivation](#motivation)
+- [Overview](#overview)
+- [Walkthrough](#walkthrough)
+- [Interface Definitions](#interface-definitions)
+- [FAQ](#faq)
+
+## Motivation
+
+Bugs abound, functions are slow, and developers need to figure out what's
+happening and why. But developers aren't typically writing WebAssembly by hand,
+they are writing in a high-level source language and compiling it down into
+WebAssembly code. Even though the WebAssembly code is what is actually being
+executed, developers prefer to debug and profile in terms of the source
+language. This source-level experience is the norm for both native code in GDB,
+LLDB, or `perf`, and for managed languages in their language-specific tooling,
+like JavaScript in a Web browser's developer tools. This proposal aims unlock
+the source-level debugging experience for Wasm code running inside a runtime.
+
+> **Note:** This proposal currently only defines APIs that are powerful enough
+> to support source-level stepping in a debugger and symbolicating source-level
+> locations and function names in other developer tools, such as a profiler or
+> logging console.
+>
+> There are many possible debugging queries that this proposal does not
+> currently define APIs for. It does not define APIs for inspecting scopes and
+> recovering their bindings' values. It does not define APIs for expanding
+> physical Wasm function frames into logical, possibly-inlined source function
+> frames. We intend to support these use cases eventually, but are starting with
+> just source-level location information, and can incrementally grow the
+> debugging module API surface as time goes on.
+
+## Overview
+
+This proposal builds on the following existing or proposed WebAssembly concepts:
+
+* **Custom sections:** These are sections defined by the [core WebAssembly
+  spec's binary
+  format](https://webassembly.github.io/spec/core/binary/modules.html#custom-section)
+  whose contents are uninterpreted by the core spec but can be interpreted by
+  other tools or specifications (including this proposal)
+* **Interface types:** The [WebAssembly interface
+  types](https://github.com/WebAssembly/interface-types/blob/master/proposals/interface-types/Explainer.md)
+  proposal allows defining shared-nothing, language-neutral interfaces.
+
+This proposal defines the following new concepts:
+
+* **Debuggee module:** A debuggee module is the Wasm module that is being
+  debugged or profiled.
+* **Debugging module:** A debugging module is referenced from its debuggee Wasm
+  module, and is a separate Wasm module that translates between Wasm-level
```

Comment:
Could the requirement that a debugging module needs to be a Wasm module be problematic? When we'll start implementing debugging modules for DWARF, for example, we could leverage existing code from LLDB/LLVM that will give us the ability of parsing DWARF files, decode DWARF information, un-mangle names according to the source language specified, supporting multiple languages, evaluate expressions to determine the value of source-level variables, and so on.
Modifying all this existing code so that it compiles to WASM could be a daunting task. It would be much simpler to have 'native' debugging module and define a (websocket-based?) debugging protocol to communicate with them.

Reply:
+1 to this. Probably we should just leave it to implementation detail.